### ISSUE TITLE
Question about usage of namespaced_function in pkg.py

### ISSUE THREAD

**Issue by hrumph** [saltstack/salt#34732]
### Would like to understand the usage of namespaced_function in pkg.py

Hi,
in pkg.py I see the following code

```
if salt.utils.is_windows():
    from salt.modules.win_pkg import _get_package_info
    from salt.modules.win_pkg import get_repo_data
    from salt.modules.win_pkg import _get_latest_pkg_version
    from salt.modules.win_pkg import _reverse_cmp_pkg_versions
    from salt.modules.win_pkg import _get_local_repo_dir
    _get_local_repo_dir = _namespaced_function(_get_local_repo_dir, globals())
    _get_package_info = _namespaced_function(_get_package_info, globals())
    get_repo_data = _namespaced_function(get_repo_data, globals())
    _get_latest_pkg_version = \
        _namespaced_function(_get_latest_pkg_version, globals())
    _reverse_cmp_pkg_versions = \
        _namespaced_function(_reverse_cmp_pkg_versions, globals())

```

Some people and myself on the efnet #salt IRC channel were asking about what _namespaced_function was accomplishing in the above code segment, and I'd like to understand this (so i can make a pull request that may depend on understanding this). @UtahDave this question is especially for you since you introduced this code.

What I would like to do is to get rid of the hard dependencies and just export the functions somehow so they can be accessed the normal way through the **salt** dictionary. For instance, I might make a function in win_pky.py called pkg_state_functions, which returns a dictionary of the private functions, which can then be accessed by the pkg.py state module. The reason why it would be nice to be rid of hard dependencies is that you can then fully replace pkg.py with an external module. Anyway before someone tells me that this is an awful and/or impossible idea, i'd still like to understand things better. THanks in advance for helping me out.

Edit:
Sorry botched original description just a little bit. I want to be able to replace win_pkg with an external module but so long as pkg.py is hard-importing bits of it, it won'y work (not if i overwrite any parts used by pkg.py).

**Comment by Ch3LL** [comment]
ping @UtahDave any way you can answer @hrumph 's questions about `_namespaced_function ` and add any clarification?

**Comment by UtahDave** [comment]
As it's name indicates, the  `_namespaced_function` imports the function into the current module's namespace.

What's happening in the code you've noted above, @hrumph, is that there are several functions in the `win_pkg` module that are needed in the `pkg` state.

The `_namespaced_function` clones the function into the current globals() namespace.

**Comment by hrumph** [comment]
Hi @UtahDave,
right, but I don't understand why these functions can't just be called in the usual manner with the \__salt\__ dictionary. Instead they're being imported creating a hard dependency. Is there some reason why that had to be done?

**Comment by UtahDave** [comment]
I think it's because the functions call eachother directly. If you look at the source of `_get_latest_pkg_version` here: https://github.com/saltstack/salt/blob/0c08724a7c39dd8e67efd764eebf644d87076fbe/salt/modules/win_pkg.py#L1218

You'll see that it calls `_reverse_cmp_pkg_versions` so they need to be in the same namespace.

I'm a bit hazy on the details here because I originally worked on that 3 years ago. @terminalmage and @s0undt3ch might have more info on that.

**Comment by hrumph** [comment]
Hi again @UtahDave. I think I understand what you are saying now. TYVM btw. I'm still newish to Python so I may be asking questions which you think should be be obvious. I'll close this in a day or to if I can't think of any followup questions.

**Comment by UtahDave** [comment]
Oh, no problem at all. You're correct that generally one should use the functions available in the `__salt__` dictionary. I think these were odd cases that required this.

**Comment by terminalmage** [comment]
``namespaced_function()`` is used to ensure that our "magic" dunder dictionaries such as ``__salt__``, ``__grains__``, etc. are available in functions imported from other modules. The reason we need to do this is because when the loader loads the functions, it adds these dunders, but this doesn't happen for functions that are imported.

The reason these functions cannot be invoked like any other function in Salt is because the loader does not add private functions (those beginning with underscores) to the ``__salt__`` dictionary. This is a conscious choice, allowing helper functions to be written within execution modules without them being callable on the CLI.

**Comment by hrumph** [comment]
Hi @terminalmage. That was a very concise and direct answer. I'll mull over this. It doesn't sound like any changes I would make would be accepted, but leave this question open for a day or two in case I have any follow up questions.

**Comment by terminalmage** [comment]
@damon-atkins I'm afraid I don't understand your question, can you rephrase it?

**Comment by damon-atkins** [comment]
I have a function _get_repo_src_dest  inside modules/win_pkg.py which needs data from __opts__  which seems to work fine. Without the need for 
```
    from salt.modules.win_pkg import _get_repo_src_dest
    _get_repo_src_dest = _namespaced_function(_get_repo_src_dest, globals())
```
How how do get Salt to error to test this is needed.

**Comment by damon-atkins** [comment]
Appears not to be required.
